File: frontend/iosApp/UseCases/Sources/UseCases/Impls/RootUseCase.swift
import Foundation
import Combine
import Repositories

public struct RootUseCase: RootUseCaseProtocol, Sendable {
    private let authSessionRepository: AuthSessionRepositoryProtocol

    public init(authSessionRepository: AuthSessionRepositoryProtocol) {
        self.authSessionRepository = authSessionRepository
    }

    public var observeDidLogout: AnyPublisher<Void, Never> {
        authSessionRepository.getSessionPublisher()
            .dropFirst()
            .filter { $0 == nil }
            .map { _ in () }
            .eraseToAnyPublisher()
    }

    public func checkPreviousSession() -> RootUseCaseModel.CheckPreviousSessionResult {
        guard let session = authSessionRepository.restore() else {
            return .notLoggedIn
        }
        return .loggedIn(session: session)
    }

    public func handleDeepLink(url: URL) -> RootUseCaseModel.HandleDeepLinkResult {
        guard let deepLink = DeepLink.parse(url: url) else {
            return .invalidURL
        }

        guard authSessionRepository.restore() != nil else {
            return .notAuthenticated(deepLink)
        }

        return .authenticated(deepLink)
    }
}

File: frontend/shared/src/commonMain/kotlin/com/example/memoriesapp/usecase/RootUseCase.kt
package com.example.memoriesapp.usecase

import com.example.memoriesapp.domain.AuthSession
import kotlinx.coroutines.flow.Flow

/**
 * Result of checking previous session
 */
sealed class CheckPreviousSessionResult {
    data class LoggedIn(val session: AuthSession) : CheckPreviousSessionResult()
    data object NotLoggedIn : CheckPreviousSessionResult()
}

/**
 * Result of handling deep link
 */
sealed class HandleDeepLinkResult {
    data class Authenticated(val deepLink: DeepLink) : HandleDeepLinkResult()
    data class NotAuthenticated(val deepLink: DeepLink) : HandleDeepLinkResult()
    data object InvalidURL : HandleDeepLinkResult()
}

interface RootUseCase {
    val observeDidLogout: Flow<Unit>
    fun checkPreviousSession(): CheckPreviousSessionResult
    fun handleDeepLink(url: String): HandleDeepLinkResult
}
